src/tokens/user_token.rs

```diff
@@ -830,6 +829,235 @@ impl ImplicitUserTokenBuilder {
+///
+/// # Examples
+///
+/// ```rust
+/// # async move {
+/// # use twitch_oauth2::{id::TwitchTokenResponse, UserToken, tokens::DeviceUserTokenBuilder, Scope};
+/// # use url::Url;
+/// # use std::borrow::Cow;
+/// # let client = twitch_oauth2::client::DummyClient; stringify!(
+/// let client = reqwest::Client::builder()
+///    .redirect(reqwest::redirect::Policy::none())
+///   .build()?;
+/// # );
+/// let mut builder = DeviceUserTokenBuilder::new("myclientid", vec![Scope::ChatRead, Scope::ChatEdit]);
+/// let code = builder.start(&client).await?;
+/// println!("Please go to {}", code.verification_uri);
+/// let token = builder.wait_for_code(&client, tokio::time::sleep).await?;
+/// println!("Token: {:?}", token);
+/// # Ok::<(), Box<dyn std::error::Error>>(())
+/// # };
+/// ```
+
+pub struct DeviceUserTokenBuilder {
+    client_id: ClientId,
+    client_secret: Option<ClientSecret>,
+    scopes: Vec<Scope>,
+    response: Option<(std::time::Instant, crate::id::DeviceCodeResponse)>,
+}
+
+impl DeviceUserTokenBuilder {
+    /// Create a [`DeviceUserTokenBuilder`]
+    pub fn new(client_id: impl Into<ClientId>, scopes: Vec<Scope>) -> Self {
+        Self {
+            client_id: client_id.into(),
+            client_secret: None,
+            scopes,
+            response: None,
+        }
+    }
+
+    /// Set the client secret, only necessary if you have one
+    pub fn set_secret(&mut self, secret: Option<ClientSecret>) { self.client_secret = secret; }
+
+    /// Get the request for getting a [`DeviceCodeResponse`](crate::id::DeviceCodeResponse)
+    pub fn get_exchange_device_code_request(&self) -> http::Request<Vec<u8>> {
+        // the equivalent of curl --location 'https://id.twitch.tv/oauth2/device' \
+        // --form 'client_id="<clientID>"' \
+        // --form 'scopes="<scopes>"'
+        use http::{HeaderMap, Method};
+        use std::collections::HashMap;
+        let mut params = HashMap::new();
+        params.insert("client_id", self.client_id.as_str());
+        let scopes = self.scopes.as_slice().join(" ");
+        if !scopes.is_empty() {
+            params.insert("scopes", &scopes);
+        }
+        crate::construct_request(
+            &crate::DEVICE_URL,
+            params,
+            HeaderMap::new(),
+            Method::POST,
+            vec![],
+        )
+    }
+
+    /// Parse the response from the device code request
+    pub fn parse_exchange_device_code_response(
+        &mut self,
+        response: http::Response<Vec<u8>>,
+    ) -> Result<&crate::id::DeviceCodeResponse, crate::RequestParseError> {
+        let response = crate::parse_response(&response)?;
+        self.response = Some((std::time::Instant::now(), response));
+        Ok(&self.response.as_ref().unwrap().1)
+    }
+
+    /// Start the device code flow
+    ///
+    /// # Notes
+    ///
+    /// Use [`DeviceCodeResponse::verification_uri`](crate::id::DeviceCodeResponse::verification_uri) to get the user token.
```

Comment:
```suggestion
    /// Use [`DeviceCodeResponse::verification_uri`](crate::id::DeviceCodeResponse::verification_uri) to get the URL the user needs to visit.
```
(or something similar)